README.md

```diff
@@ -1334,6 +1337,156 @@ CREATE VIEW animal AS
+```graphql
+{
+    Animal @macro_edge_definition(name: "out_Animal_GrandparentOf") {
+        out_Animal_ParentOf {
+            out_Animal_ParentOf @macro_edge_target {
+                uuid
+            }
+        }
+    }
+}
+```
+The directive `@macro_edge_definition` declares that this is a macro, and gives it a name.
+To use it as any other edge, we will refer to it by the name given here.
+The directive `@macro_edge_target` specifies where the grandchild of the animal is.
+Now let's use this macro to rewrite our query:
+```graphql
+{
+    Animal {
+        name @output(out_name: "animal_name")
+        out_Animal_GrandparentOf {
+            name @output(out_name: "grandchild_name")
+        }
+    }
+}
+```
+Macro edges can make use of tags, filters, etc, to define more complicated macros like this one
+```graphql
+Animal @macro_edge_definition(name: "out_Animal_RichYoungerSiblings") {
+    net_worth @tag(tag_name: "net_worth")
+    out_Animal_BornAt {
+        event_date @tag(tag_name: "birthday")
+    }
+    in_Animal_ParentOf {
+        out_Animal_ParentOf @macro_edge_target {
+            net_worth @filter(op_name: ">", value: ["%net_worth"])
+            out_Animal_BornAt {
+                event_date @filter(op_name: "<", value: ["%birthday"])
+            }
+        }
+    }
+}
+```
+Macro edges can also use runtime parameters. The values for the parameters are provided when
+the macro is registered, and not when it is used. See the section on
+[using the macro registry](#using-the-macro-registry) for an example of how to use macros
+with runtime parameters.
+
+Macro definitions and macro uses are type checked. If a macro is invalid or is used incorrectly,
+the macro system will complain, rather than create an invalid query.
+
+#### Constraints for macro definitions
+- Macro definitions cannot use other macros as part of their definition.
+- A macro requires exactly one `@macro_edge_definition` and one `@macro_edge_target` directive.
+Additionally, these directives can only be used inside macro definitions.
+- The `@macro_edge_target` cannot be inside a `@fold` scope.
+- The `@macro_edge_target` cannot begin with a type coercion. If that is what you want, put
+the `@macro_edge_target` on the type coercion itself, instead of before it, like this:
+`in_Entity_Related { ... on Animal @macro_edge_target {uuid} }`.
+- Macros cannot use `@output` or `@output_source`.
+
+#### Constraints for macro usage
+- The `@optional` and `@recurse` directives cannot be used on macro edges.
+
+### Using the macro registry
+This example demonstrates how to define a macro and use it. It assumes you already know
+how to construct a GraphQL schema.
+
+```python
+from graphql_compiler import macros
+from graphql_compiler.compiler.subclass import compute_subclass_sets
+
+def example(schema, type_equivalence_hints):
+    # Create a macro registry based on the schema
+    subclass_sets = compute_subclass_sets(schema, type_equivalence_hints)
```

Comment:
Since `compute_subclass_sets` is O(n^3) function, I don't think that we should include it in the README.md since it will take too long for large schemas. So either optimize `compute_subclass_sets` or use the SchemaGraph. You can optimize the `compute_subclass_sets` by using some of the code that we already have for the SchemaGraph generation though.